tests/0122-buffer_cleaning_after_rebalance.c

```diff
@@ -0,0 +1,226 @@
+
+        switch (err) {
+        case RD_KAFKA_RESP_ERR__ASSIGN_PARTITIONS:
+                TEST_SAY("Group rebalanced (%s): "
+                         "%d new partition(s) assigned\n",
+                         rd_kafka_rebalance_protocol(rk), partitions->cnt);
+
+                if (!strcmp(rd_kafka_rebalance_protocol(rk), "COOPERATIVE"))
+                        error = rd_kafka_incremental_assign(rk, partitions);
+                else
+                        ret_err = rd_kafka_assign(rk, partitions);
+
+                break;
+        case RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS:
+                TEST_SAY("(%s): %d partition(s) revoked\n",
+                         rd_kafka_rebalance_protocol(rk), partitions->cnt);
+
+                if (!strcmp(rd_kafka_rebalance_protocol(rk), "COOPERATIVE"))
+                        error = rd_kafka_incremental_unassign(rk, partitions);
+                else
+                        ret_err = rd_kafka_assign(rk, NULL);
+                break;
+
+        default:
+                break;
+        }
+
+        if (error) {
+                TEST_FAIL("%% incremental assign failure: %s\n",
+                          rd_kafka_error_string(error));
+                rd_kafka_error_destroy(error);
+        } else if (ret_err) {
+                TEST_FAIL( "%% assign failure: %s\n",
+                rd_kafka_err2str(ret_err));
+        }
+}
+
+/**
+ * Consume with batch + queue interface
+ *
+ */
+static int do_test_consume_batch (char *strategy) {
+        int partition_cnt = 4;
+        rd_kafka_queue_t *rkq1, *rkq2;
+        int producemsgcnt = 400;
+        const char *topic;
+        rd_kafka_t *c2;
+        int p;
+        int timeout_ms = 30000;
+        uint64_t testid;
+        int consumemsgcnt = 500;
+        rd_kafka_conf_t *conf;
+        thrd_t thread_id;
+        thrd_t thread_id2;
+        struct consumer_s c1_args;
+        struct consumer_s c2_args;
+        test_conf_init(&conf, NULL, 60);
+        test_conf_set(conf, "enable.auto.commit", "false");
+        test_conf_set(conf, "auto.offset.reset", "earliest");
+
+        testid = test_id_generate();
+
+        /* Produce messages */
+        topic = test_mk_topic_name(__FUNCTION__, 1);
+
+        for (p = 0 ; p < partition_cnt ; p++)
+                test_produce_msgs_easy(topic,
+                                       testid,
+                                       p,
+                                       producemsgcnt / partition_cnt);
+        /* Create simple consumer */
+        if(strcmp(strategy, "cooperative-sticky") == 0)
+                test_conf_set(conf, "partition.assignment.strategy", "cooperative-sticky");
+
+        rd_kafka_conf_set_rebalance_cb(conf, rebalance_cb);
+
+        c1 = test_create_consumer(topic, rebalance_cb,
+                                  rd_kafka_conf_dup(conf), NULL);
+        c2 = test_create_consumer(topic, NULL, conf, NULL);
+
+        test_consumer_subscribe(c1, topic);
```

Comment:
Even with the updated code I think there's a problem:
1. c1 subscribes
2. wait for c1 assignment
3. c2 subscribes
4. ..rebalance
5. wait for c2 assignment
6. then consume c1 and c2

Problem is that since no messages are beginning to be batched in a consume_batch..() call there is no messages to clear out when the barrier is received, those messages will already be purged by "outdated()" further up in the q_serve loop.

Did you attempt to run this test-case with a non-fixed version of librdkafka to verify that the test-case triggers the original bug?